doc/design_AsyncNDRegisterAcessor_and_NumericAddressedBackend.dox

```diff
@@ -0,0 +1,60 @@
+namespace ChimeraTK {
+/**
+\page design_AsyncNDRegisterAccessor_NumericAddressed Design: AsyncNDRegisterAccessor and the asynchronous implelentation in the NumericAddressedBackend
+
+\section  design_AsyncNDRegisterAccessor The AsyncNDRegisterAccessor
+
+The AsyncNDRegisterAccessor is an extension of the NDRegisterAccessor for push-type data.
+
+The NDRegisterAccessor itself extends the TransferElement with a user-type buffer. The TransferElement does not have a user type buffer and serves as a base class such that implementations with different user types can be stored together in containers. The TransferElement also contais a notification queue to implement the asynchronous read functionality. This notification queue is of type void (notifications only, no data is transported) because the data type is not known. The AsyncNDRegisterAccessor now povides the data transport queue which is needed for asynchronous accessors with data. As this queue is comparatively resource-intensive (multiple instances of the user buffer object pre-allocated in each queue) and not needed for synchronous data transfer, it is not directly added in the NDRegisterAccessor but in a second class.
+
+The data transport queue is implmented as continuation of the notification queue, such that the read operations of the notification queue trigger the filling of the user buffer.
+
+The AsyncNDRegisterAccessor is ready to use. It provides a generic asynchronous implementation.
+Each backend has to store a list of accessors and call AsyncNDRegisterAccessor::activate(), AsyncNDRegisterAccessor::deactivate(), AsyncNDRegisterAccessor::sendDestructively() and AsyncNDRegisterAccessor::sendException() appropriately. As the way the data containers are organised and the asynchronous mechanism is implemented stongly depend on the backend, the implementation of these are left to the specific backend.
+However, there is a common interface implemented to allow the clean subscription/unsubscription from the backend. The AsyncNDRegisterAccessor is templated to an AsyncAccessorManager and an AsyncAccessorID, which are both handed over in the constructor. In its destructor the AsyncNDRegisterAccessor is calling
+<pre>
+void AsyncAccessorManager::unsubscribe(AsyncAccessorID id);
+</pre>
+This function has to be implemented by the manager. From the ID it knows which subscriber is to be removed. For this mechanism to work, the AsyncAccessorManager must store boost::weak_pointers and lock them when the instance is being accessed. Otherwise the one instance held within the manager prevents the destructor from being called.
```

Comment:
The structure is a but unclear to me. It looks like one has to use a manager (because the accessor is templated to it). Before you state that the functions `activate()` etc. need to be called by the backend. This is confusing. They should be called by the manager, right?

Reply:
Also I am wondering if we could (maybe later) split a generic portion off the NumericAddressedInterruptDispatcher into an AsyncAccessorManager base class, which would not be mandatory to use but likely used by most backends. This could deal with maintaining the subscriptions and copying the incoming data and passing it to all subscribers, so the backend has to call merely one function when new data arrives.

Reply:
I had the same idea. But the way it is done in the NumericAddressedBackend is already very special. The actual list of weak pointers is inside of a helper object AsyncVariable, which contains all the subscribers for one particular accessor, together with the according synchronous accessor to read the data. That is the first list which changes with a subscription/unsubscription.
The dispatcher then holds a list of those AsyncVariables, which can also grow/shrink. All those lists have to be protected with a lock against the dispatcher thread.
One would only benefit if the subscribe/unsubscribe would be in the generic manager, but I did not see how to fit this without the special behaviour of one async variable which feeds a sub-set. The Doocs-bakend would have a completely different mechanism.

Reply:
~~The Doocs-bakend would have a completely different mechanism.~~
Scratch that. It's exactly the same. Multiple subscribers to the same variable must be handled, and there is a list of variables.

Reply:
I will see if I can already separate that in a clever way. But the current implementation has a lock. I don't know if this is acceptable for a generic solution.

For the AsyncNDRegisterAccessor it does not make a difference. But having a base class which does the fiddly stuff and has some rules is a good idea